src/webgpu/api/operation/command_buffer/render/state_tracking.spec.ts

```diff
@@ -179,3 +184,98 @@ g.test('set_index_buffer_without_changing_buffer')
       );
     }
   });
+
+g.test('set_vertex_buffer_without_changing_buffer')
+  .desc(
+    `
+  Test that setting vertex buffer states (offset, size) multiple times in different orders still
+  keeps the correctness of each draw call.
+`
+  )
+  .params(u =>
+    u.combine('drawParams', [
+      [
+        // Change 'size' then 'offset' in setVertexBuffer()
+        { offset: 0, vertices: 1 },
+        { offset: 0, vertices: 2 },
+        { offset: 2, vertices: 2 },
+      ],
+      [
+        // Change 'offset' then 'size' in setVertexBuffer()
+        { offset: 0, vertices: 1 },
+        { offset: 1, vertices: 1 },
+        { offset: 1, vertices: 3 },
+      ],
+    ] as const)
+  )
+  .fn(async t => {
+    const { drawParams } = t.params;
+
+    const kPositions = [-0.75, -0.25, 0.25, 0.75];
+    const kColors = [
+      new Uint8Array([255, 0, 0, 255]),
+      new Uint8Array([0, 255, 0, 255]),
+      new Uint8Array([0, 0, 255, 255]),
+      new Uint8Array([255, 0, 255, 255]),
+    ];
+
+    // Initialize the vertex buffer with required vertex attributes (position: f32, color: f32x4)
+    const kVertexAttributesCount = 4;
+    const vertexBuffer = t.device.createBuffer({
+      usage: GPUBufferUsage.VERTEX,
+      size: t.kVertexAttributeSize * kVertexAttributesCount,
+      mappedAtCreation: true,
+    });
+    t.trackForCleanup(vertexBuffer);
+    const vertexAttributes = vertexBuffer.getMappedRange();
+    for (let i = 0; i < kPositions.length; ++i) {
+      const baseOffset = t.kVertexAttributeSize * i;
+      const vertexPosition = new Float32Array(vertexAttributes, baseOffset, 1);
+      vertexPosition[0] = kPositions[i];
+      const vertexColor = new Uint8Array(vertexAttributes, baseOffset + 4, 4);
+      vertexColor.set(kColors[i]);
+    }
+
+    vertexBuffer.unmap();
+
+    const renderPipeline = t.GetRenderPipelineForTest();
+
+    const outputTexture = t.device.createTexture({
+      format: 'rgba8unorm',
+      size: [kPositions.length, 1, 1],
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: outputTexture.createView(),
+          loadValue: [0, 0, 0, 1],
+          storeOp: 'store',
+        },
+      ],
+    });
+    renderPass.setPipeline(renderPipeline);
+
+    for (const drawParam of drawParams) {
+      renderPass.setVertexBuffer(
+        0,
+        vertexBuffer,
+        drawParam.offset * t.kVertexAttributeSize,
+        drawParam.vertices * t.kVertexAttributeSize
+      );
+      renderPass.draw(drawParam.vertices);
```

Comment:
There should be tests for multiple setVertexBuffer calls in a row before draw. This would improve coverage since draw() will necessarily flush any dirty bits and because this isn't a case you would normally see in applications so bugs could evade detection.

Maybe the drawParams should just be hardcoded setVertexBuffer/draw calls in the test function, instead of parameterized; that way you don't have to figure out how to parameterize things like "setVertexBuffer, setVertexBuffer, draw".

Reply:
I've updated the code here. PTAL, thanks!